discover.py

```diff
@@ -33,28 +34,68 @@ def workers() -> Dict[WorkerName, List[Task]]:
     }
 
 
-def tasks() -> Dict[TaskType, List[TaskName]]:
+def tasks(all_workers: Dict[WorkerName, List[Task]]=None) \
+        -> Dict[TaskType, List[TaskName]]:
     """Find tasks available in the cluster
 
     Returns:
         Dictionary of task types with their algorithm.
     """
-    all_tasks = list(chain.from_iterable(workers().values()))
+    if all_workers is None:
+        all_workers = workers()
+    all_tasks = list(chain.from_iterable(all_workers.values()))
     all_tasks = sorted(all_tasks, key=attrgetter('type'))
     return {
         task_type: {task.name for task in unitype_tasks}
         for task_type, unitype_tasks in groupby(all_tasks, key=attrgetter('type'))
     }
 
 
-def role(task_name: TaskName) -> TaskType:
+def role(task_name: TaskName,
+         all_tasks: Dict[TaskType, List[TaskName]]=None) -> TaskType:
     """Find type of the task.
 
     Returns:
         Type of the task.
     """
-    all_tasks = tasks()
+    if all_tasks is None:
+        all_tasks = tasks()
     for task_type in all_tasks:
         if task_name in all_tasks[task_type]:
             return task_type
     raise KeyError(task_name)
```

Comment:
I consider this as a lookup into a set of tasks, what it in fact is. In such context `KeyError` seems reasonable.

Reply:
This is because you know the implementation. From the point of API user, it is the incorrect value that is important, not that some key lookup failed.

Reply:
Debatable, but accepted.